api/v1/api_test.go

```diff
@@ -0,0 +1,273 @@
+	if len(mainMap) != len(testMap) {
+		return false
+	}
+
+	for key, _ := range mainMap {
+		if mainMap[key].(string) != testMap[key] {
+			return false
+		}
+	}
+	return true
+}
+
+func compareArray(a, b []byte) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for i, v := range a {
+		if v != b[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func TestStatusAPI(t *testing.T) {
+	dms := storage.NewDiskMetricStore("", 100*time.Millisecond, nil, logger)
+	TestApi := New(logger, dms, testFlags, testBuildInfo)
+
+	req, err := http.NewRequest("GET", "http://example.org/", &bytes.Buffer{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	w := httptest.NewRecorder()
+
+	testResponse := response{}
+	TestApi.status(w, req)
+	json.Unmarshal(w.Body.Bytes(), &testResponse)
+	jsonData := testResponse.Data.(map[string]interface{})
+	responseFlagData := jsonData["flags"].(map[string]interface{})
+	responseBuildInfo := jsonData["build_information"].(map[string]interface{})
+
+	if expected, got := http.StatusOK, w.Code; expected != got {
+		t.Errorf("Wanted status code %v, got %v.", expected, got)
+	}
+
+	if !compareMaps(responseFlagData, testFlags) {
+		t.Errorf("Wanted following flags %v, got %v.", testFlags, responseFlagData)
+	}
+
+	if !compareMaps(responseBuildInfo, testBuildInfo) {
+		t.Errorf("Wanted following build info %v, got %v.", testBuildInfo, responseBuildInfo)
+	}
+}
+
+func TestMetricsAPI(t *testing.T) {
+	dms := storage.NewDiskMetricStore("", 100*time.Millisecond, nil, logger)
+	TestApi := New(logger, dms, testFlags, testBuildInfo)
+
+	req, err := http.NewRequest("GET", "http://example.org/", &bytes.Buffer{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	w := httptest.NewRecorder()
+
+	TestApi.metrics(w, req)
+
+	if expected, got := http.StatusOK, w.Code; expected != got {
+		t.Errorf("Wanted status code %v, got %v.", expected, got)
+	}
+
+	requiredResponse, _ := json.Marshal(&response{
+		Status: statusSuccess,
+		Data:   []interface{}{},
+	})
+
+	if !compareArray(requiredResponse, w.Body.Bytes()) {
+		t.Errorf("Wanted response %v, got %v.", string(requiredResponse), w.Body.String())
+	}
+
+	testTime := time.Now()
+
+	errCh := make(chan error, 1)
+
+	dms.SubmitWriteRequest(storage.WriteRequest{
+		Labels:         grouping1,
+		Timestamp:      testTime,
+		MetricFamilies: metricFamiliesMap(mf1, mf2),
+		Done:           errCh,
+	})
+
+	for err := range errCh {
+		t.Fatal("Unexpected error:", err)
+	}
+
+	w = httptest.NewRecorder()
+
+	TestApi.metrics(w, req)
+
+	requiredResponse, _ = json.Marshal(&response{
+		Status: statusSuccess,
+		Data:   templateMetrics(dms),
+	})
+
+	if expected, got := http.StatusOK, w.Code; expected != got {
+		t.Errorf("Wanted status code %v, got %v.", expected, got)
+	}
+
+	if !compareArray(requiredResponse, w.Body.Bytes()) {
+		t.Errorf("Wanted response %v, got %v.", string(requiredResponse), w.Body.String())
+	}
+}
+
+func metricFamiliesMap(mfs ...*dto.MetricFamily) map[string]*dto.MetricFamily {
```

Comment:
I'd also copy the doc comment from where you copied this (diskmetricstore_test.go) because it is otherwise confusing to the reader why you are doing this.

Having said that, as this gets close to a significant amount of not entirely trivial code, perhaps we should move it to a `testutil` package (or perhaps even `internal/testutil`, see https://dave.cheney.net/2019/10/06/use-internal-packages-to-reduce-your-public-api-surface ).